packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts

```diff
@@ -0,0 +1,121 @@
+   *
+   * @default "Allows CloudFront to reach to the bucket"
+   */
+  readonly comment?: string;
+}
+
+/**
+ * Interface for CloudFront OriginAccessIdentity
+ */
+export interface IOriginAccessIdentity extends cdk.IResource, iam.IGrantable  {
+  /**
+   * The Origin Access Identity Name
+   */
+  readonly originAccessIdentityName: string;
+}
+
+abstract class BaseOriginAccessIdentity extends cdk.Resource {
+  /**
+   * The Origin Access Identity Name (physical id)
+   */
+  public abstract readonly originAccessIdentityName: string;
+  /**
+   * Derived principal value for bucket access
+   */
+  public abstract readonly grantPrincipal: iam.IPrincipal;
+
+  /**
+   * The ARN to include in S3 bucket policy to allow CloudFront access
+   */
+  protected arn(): string {
+    return cdk.Stack.of(this).formatArn(
+      {
+        service: "iam",
+        region: "", // global
+        account: "cloudfront",
+        resource: "user",
+        resourceName: `CloudFront Origin Access Identity ${this.originAccessIdentityName}`
+      }
+    );
+  }
+}
+
+/**
+ * An origin access identity is a special CloudFront user that you can
+ * associate with Amazon S3 origins, so that you can secure all or just some of
+ * your Amazon S3 content.
+ *
+ * @resource AWS::CloudFront::CloudFrontOriginAccessIdentity
+ */
+export class OriginAccessIdentity extends BaseOriginAccessIdentity implements IOriginAccessIdentity {
+  /**
+   * Creates a OriginAccessIdentity by providing the OriginAccessIdentityName
+   */
+  public static fromOriginAccessIdentityName(
+    scope: cdk.Construct,
+    id: string,
+    originAccessIdentityName: string): IOriginAccessIdentity {
+
+    class Import extends BaseOriginAccessIdentity {
+      public readonly originAccessIdentityName = originAccessIdentityName;
+      public readonly grantPrincipal = new iam.ArnPrincipal(this.arn());
+    }
+
+    return new Import(scope, id);
+  }
+
+  /**
+   * CDK L1 resource
+   */
+  public readonly resource: CfnCloudFrontOriginAccessIdentity;
+
+  /**
+   * The Amazon S3 canonical user ID for the origin access identity, used when
+   * giving the origin access identity read permission to an object in Amazon
+   * S3.
+   *
+   * @attribute S3CanonicalUserId
+   */
+  public readonly s3CanonicalUserId: string;
+
+  /**
+   * Derived principal value for bucket access
+   */
+  public readonly grantPrincipal: iam.IPrincipal;
+
+  /**
+   * The Origin Access Identity Name (physical id)
+   */
+  public readonly originAccessIdentityName: string;
+
+  constructor(scope: cdk.Construct, id: string, props?: OriginAccessIdentityProps) {
+    super(scope, id);
+
+    this.resource = new CfnCloudFrontOriginAccessIdentity(this, "Resource", {
+      cloudFrontOriginAccessIdentityConfig: {
+        comment: (props && props.comment) || "Allows CloudFront to reach the bucket"
+      }
+    });
+    // physical id - OAI name
+    this.originAccessIdentityName = this.resource.ref;
```

Comment:
please apply our pattern for physical IDs. 

- call super constructor with `physicalName` ([example](https://github.com/aws/aws-cdk/blob/652a8f5b4d253babb57dfb3f900794071509a565/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L993))
- use `this.physicalName` when passed to L1 ([example](https://github.com/aws/aws-cdk/blob/652a8f5b4d253babb57dfb3f900794071509a565/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L1001)
- use `this.getResourceNameAttribute` and `this.getResourceArnAttribute` to render the name and ARN attributes ([example](https://github.com/aws/aws-cdk/blob/652a8f5b4d253babb57dfb3f900794071509a565/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L1017))./

I am aware this is not documented anywhere and we haven't found a way to enforce it.

Reply:
Thanks for the help. The `physicalName` is the origin access id which is returned when creating OAI resource. It is not known before.

> 1) call super constructor with `physicalName`

The constructor does not get any physicalName so I can't pass it to `super`. But I added it at least to the `Import` class.

> 2) use `this.physicalName` when passed to L1

The `physicalName` is never passed to the L1 because it's not known before.

> 3) use `this.getResourceNameAttribute` and `this.getResourceArnAttribute` to render the name and ARN attributes

I used the `getResourceNameAttribute` but since OAIs don't have ARNs I can't use `getResourceArnAttribute`.